src/AggregateFunctions/AggregateFunctionFunnelAnalysis.h

```diff
@@ -0,0 +1,288 @@
+        TimestampEvents & left = doFinalMerge(events, begin, mid);
+        TimestampEvents & right = doFinalMerge(events, mid+1, end);
+        const auto size = left.size();
+        left.insert(std::begin(right), std::end(right));
+        const auto b = std::begin(left);
+        const auto m = std::next(b, size);
+        const auto e = std::end(left);
+
+        std::inplace_merge(b, m, e, Comparator{});
+        return left;
+    }
+
+    void serialize(WriteBuffer & buf) const
+    {
+        writeBinary(events_lists_to_merge.size(), buf);
+        for (const auto & events : events_lists_to_merge)
+        {
+            writeBinary(events.size(), buf);
+            for (const auto & event : events)
+            {
+                writeBinary(event.first, buf);
+                writeBinary(event.second, buf);
+            }
+        }
+    }
+
+    void deserialize(ReadBuffer & buf)
+    {
+        size_t size = 0;
+        size_t vector_size;
+        readBinary(vector_size, buf);
+
+        /// TODO Protection against huge size
+        T timestamp;
+        UInt8 event;
+
+        for (size_t i = 0; i < vector_size; ++i)
+        {
+            size_t array_size;
+            readBinary(array_size, buf);
+            TimestampEvents events(array_size);
+            for (size_t j = 0; j < array_size; ++j)
+            {
+                readBinary(timestamp, buf);
+                readBinary(event, buf);
+                events.emplace_back(timestamp, event);
+                size++;
+            }
+            events_lists_to_merge.emplace_back(std::begin(events), std::end(events));
+        }
+        _size = size;
+    }
+};
+
+/** Calculates the max event level in a sliding window.
+  * The max size of events is 32, that's enough for funnel analytics
+  * T : the type of timestamp.
+  * Data : the type of the aggregate data.
+  *
+  * Usage:
+  * - funnelAnalysis(window, strict, strict_order, event1, event2, event3, ....)(event, timestamps)
+  */
+template <typename T, typename Data>
+class AggregateFunctionFunnelAnalysis final
+    : public IAggregateFunctionDataHelper<Data, AggregateFunctionFunnelAnalysis<T, Data>>
+{
+private:
+    using IndexByName = std::unordered_map<String, size_t>;
+
+    UInt64 window;
+    UInt8 events_size;
+    UInt8 strict;   // When the 'strict' is set, it applies conditions only for the not repeating values.
+    UInt8 strict_order; // When the 'strict_order' is set, it doesn't allow interventions of other events.
+                        // In the case of 'A->B->D->C', it stops finding 'A->B->C' at the 'D' and the max event level is 2.
+    IndexByName events_index;
+
+    // Loop through the entire events_list, update the event timestamp value
+    // The level path must be 1---2---3---...---check_events_size, find the max event level that satisfied the path in the sliding window.
+    // If found, returns the max event level, else return 0.
+    // The Algorithm complexity is O(n).
```

Comment:
In case we make `finalMerge` here, complexity is O(nlogk) if I am right. Need a comment about it.